File: exercises/exercism/python/circular-buffer/circular_buffer.py
class BufferFullException(Exception):
    pass


class BufferEmptyException(Exception):
    pass


class CircularBuffer(object):
    def __init__(self, capacity):
        self.buffer = [0] * capacity
        self.size = 0
        self.capacity = capacity
        self.head_index = 0
        self.tail_index = 0

    def read(self):
        if self.size == 0:
            raise BufferEmptyException("Buffer is empty. Cannot read empty buffers. Add elements to buffer.")

        data = self.buffer[self.head_index]
        self.head_index = (self.head_index + 1) % self.capacity
        self.size -= 1

        return data


    def write(self, data):

        if self.size == self.capacity:
            raise BufferFullException("Buffer is full. Remove elements from buffer to add new elements.")

        self.buffer[self.tail_index] = data
        self.tail_index = (self.tail_index + 1) % self.capacity
        self.size += 1



    def overwrite(self, data):
        if self.size != self.capacity:
            self.write(data)
        else:
            self.buffer[self.head_index % self.capacity] = data
            self.head_index = (self.head_index + 1) % self.capacity


    def clear(self):
        self.buffer = [0] * self.capacity
        self.size = 0
        self.head_index = 0
        self.tail_index = 0

File: exercises/exercism/java/circular-buffer/src/main/java/CircularBuffer.java
import java.util.ArrayList;

public class CircularBuffer<Integer> {


	private int capacity;
	private int head;
	private int tail;
	private int data[];
	
	public CircularBuffer(int capacity) throws BufferIOException {
		this.capacity = capacity;
		head = 0;
		tail = 0;
		data = new int[capacity];
	}
	
	public void write(int value) throws BufferIOException {
		
		if (isBufferFull()) {
			throw new ExpectedException("Tried to write to full buffer");
		}
		
		data[tail++] = value;
		
		tail++;
		if (tail == data.length) tail = 0;
	}
	
	public int read() throws BufferIOException {
		
		if (isBufferEmpty())
			throw new ExpectedException("Tried to read from empty buffer");
		

		int value = data[head++];
		
		head++;
		if (head == data.length) head = 0;
		
		return value;
	}
	
	public void overwrite(int value) throws BufferIOException {
		if (tail <= capacity) write(value);
		else data[head] = value;
		
		head++;
		
	}
	
	public void clear() {
		head = 0; tail = 0;
		
	}
	
	private boolean isBufferFull() {
		if (tail + 1 == head) return true;
		if (tail == (data.length - 1) && head == 0) return true;
		
		return false;
	}
	
	private boolean isBufferEmpty() {
		return head == 0 && tail == 0;
	}
}
